src/content/blog/2024-04-10-the-telemetry-tale.md

```diff
@@ -0,0 +1,117 @@
+
+### Architecture and Instrumentation
+
+![The Rafiki telemetry architecture](/developers/img/blog/2024-04-10/telemetry-architecture.png)
+
+The [number of transactions](https://github.com/interledger/rafiki/blob/d3be6b8d151d8cebc32b862e52a7bb678674d48e/packages/backend/src/open_payments/payment/outgoing/lifecycle.ts#L84-L90) is extracted from the Open Payments outgoing payment lifecycle and the value metric is handled by a [telemetry middleware](https://github.com/interledger/rafiki/blob/d3be6b8d151d8cebc32b862e52a7bb678674d48e/packages/backend/src/payment-method/ilp/connector/core/middleware/telemetry.ts) layer inside the ILP connector core.
+
+In summary, before these metrics are sent to the Otel collectors we’ve implemented local privacy-preserving measures, including:
+- packetization which is inherent in the Interledger Protocol (ILP)
+- a currency conversion service that uses a custom AWS Lambda function that pulls daily exchange rates from configured sources and stores them in an Amazon S3 bucket
+- rounding data to a precision point determined by bucketing
+- local differential privacy in the form of adding Laplacian noise to anonymize transaction data
+
+Finally, we grappled with the dilemma of disentangling playground data from our test network and live production data. The [Test Network](https://github.com/interledger/testnet) (Testnet) is an open Interledger network using test money within example wallet and e-commerce applications. Rafiki also provides a [local playground](https://github.com/interledger/rafiki/tree/main/localenv) to experiment with. We’d like to monitor activity within the testing environments to gauge interest in the Interledger Protocol (ILP) and identify usage patterns by new participants. We’ve opted to provide users the choice between the test network (testnet) and the live network (livenet). To support this dual-environment approach, the infrastructure for telemetry was effectively expanded to include two separate services: one dedicated to collecting and managing telemetry data from the test network and another for the live network. This setup ensures that data from each environment is handled independently.
+
+We’re using an AWS ECS Fargate deployment for our Telemetry Cluster. The cluster holds services for our two environments: livenet and testnet, both of them having a Network Load Balancer (NLB) in order to load-balance over their own set of ECS task replicas. By using an NLB instead of the Application Load Balancer (ALB), we are not taking the hit of parsing the L7 ([OSI Model](https://en.wikipedia.org/wiki/OSI_model)) protocols and we do not have to do TLS termination at load balancer level. In our case, we just need to pass the data received to Prometheus, and application-layer routing decisions are not needed. Acting on L4 is of greater convenience, as all we need is the TCP packets being redirected to one of the running ECS tasks.
```

Comment:
```suggestion
**Understanding Terminology**

In our telemetry deployment, it's important to understand several key terms:

- **Cluster**: This is the foundation of our ECS deployment. A cluster is a group of virtual or physical machines (known as container instances) that hosts the containerized applications.

- **Service**: Each service represents a distinct component of our telemetry infrastructure. In ECS, a service is responsible for defining how tasks (instances of containers) are run and maintained. Services are created based on ECS Task Definitions.

- **ECS Task**: A task is the smallest unit of deployment in ECS. It represents a single instance of a container running on an ECS cluster.

- **ECS Task Definition**: This is a blueprint that defines how an ECS task should be configured and run. It includes specifications such as which container images to use, CPU and memory requirements, networking configuration, and more.

- **ADOT Collector**: An ADOT (Amazon Distributed OpenTelemetry) collector is a component responsible for collecting telemetry data from various sources, such as applications and infrastructure services. It standardizes and exports this data to a monitoring backend for further analysis and visualization.

## How We Use the Above Mentioned

Within our telemetry deployment, our ECS cluster hosts two distinct services: Livenet Service and Testnet Service. Each service comprises one or more ECS tasks running on the cluster. These tasks are defined by their respective ECS task definitions, specifying them as ADOT collector containers. Notably, these containers are customized to route data to either the livenet or testnet Prometheus instance, ensuring data segregation based on environment.

Each service is associated with its own Network Load Balancer (NLB), which ensures that incoming traffic is distributed among the ECS tasks within the service.

## Why Network Load Balancer (NLB)?

We've opted for NLB over the Application Load Balancer (ALB) for several reasons:

- **Layer 4 (L4) Load Balancing**: NLB operates at the transport layer of the OSI model, meaning it focuses solely on routing traffic based on IP addresses and TCP/UDP ports. This enables efficient handling of TCP packets without the need to interpret the content of the packets, which reduces computational overhead and improves performance.

- **TLS Termination**: Unlike ALB, NLB does not require TLS termination. This means that the load balancer does not need a certificate for decrypting incoming TLS traffic. Since our objective is solely to forward encrypted traffic without making routing decisions based on its content, bypassing TLS termination simplifies the process by eliminating the need for decryption and subsequent re-encryption at the load balancer. This streamlined approach ensures efficient transmission of encrypted data to the backend services without additional processing overhead.

```

Reply:
Thanks Beni but I worry this explanation is very long and detailed. If we need to really dig into this it almost feels like we could consider a follow up blog post?

Reply:
```suggestion
Our telemetry cluster is deployed on AWS, utilizing the ECS Fargate service to support two distinct operational environments: the live network environment (livenet) and the test network environment (testnet). Since the two environments are separate, it is the Rafiki client that determines where to send the data. This way each environment has its own load balancer to handle incoming traffic and the load balancers themselves do not need to decrypt HTTPS traffic to determine where it needs to go. This allows us to establish a TLS secure channel directly with the services behind the load balancers (the respective Prometheus instances), rather than with a load balancer itself. This configuration optimizes our resource utilization, leading to a more efficient and cost-effective telemetry infrastructure. 
```